File: backend/app/models/motion.py
"""Data models for motion graphics projects."""

from __future__ import annotations

from enum import Enum
from typing import Any, Optional, Union

from pydantic import BaseModel, Field


# ─── Easing ─────────────────────────────────────────────────────────────────

class EasingName(str, Enum):
    linear = "linear"
    ease_in = "easeIn"
    ease_out = "easeOut"
    ease_in_out = "easeInOut"
    spring = "spring"
    bounce = "bounce"


EasingType = Union[EasingName, list[float]]  # named or cubic-bezier [x1,y1,x2,y2]


# ─── Keyframe ───────────────────────────────────────────────────────────────

class Keyframe(BaseModel):
    id: str
    time: float  # seconds
    value: Union[float, str]
    easing: EasingType = EasingName.ease_in_out


# ─── Layer Properties ───────────────────────────────────────────────────────

class LayerTransform(BaseModel):
    x: float = 0
    y: float = 0
    width: float = 200
    height: float = 200
    rotation: float = 0
    opacity: float = 1
    scaleX: float = 1
    scaleY: float = 1
    anchorX: float = 0.5
    anchorY: float = 0.5


class LayerAppearance(BaseModel):
    fill: str = "#ffffff"
    stroke: str = "transparent"
    strokeWidth: float = 0
    borderRadius: float = 0


class LayerEffects(BaseModel):
    blur: float = 0
    shadowX: float = 0
    shadowY: float = 0
    shadowBlur: float = 0
    shadowColor: str = "rgba(0,0,0,0.3)"
    shadowOpacity: float = 0


class TextProperties(BaseModel):
    text: str = "Text"
    fontSize: float = 48
    fontFamily: str = "Inter"
    fontWeight: int = 600
    letterSpacing: float = 0
    lineHeight: float = 1.2
    textAlign: str = "center"
    color: str = "#ffffff"


class ShapeProperties(BaseModel):
    shapeType: str = "rectangle"  # rectangle, ellipse, polygon, line, star
    points: Optional[int] = None


class PathProperties(BaseModel):
    trimStart: float = 0
    trimEnd: float = 1


# ─── Layer Types ────────────────────────────────────────────────────────────

class LayerType(str, Enum):
    text = "text"
    shape = "shape"
    image = "image"
    video = "video"
    group = "group"
    svg = "svg"


class MotionLayer(BaseModel):
    id: str
    name: str
    type: LayerType
    visible: bool = True
    locked: bool = False
    inPoint: float = 0
    outPoint: float = 5
    transform: LayerTransform = Field(default_factory=LayerTransform)
    appearance: LayerAppearance = Field(default_factory=LayerAppearance)
    effects: LayerEffects = Field(default_factory=LayerEffects)
    textProps: Optional[TextProperties] = None
    shapeProps: Optional[ShapeProperties] = None
    pathProps: Optional[PathProperties] = None
    imageSrc: Optional[str] = None
    videoSrc: Optional[str] = None
    keyframes: dict[str, list[Keyframe]] = Field(default_factory=dict)
    children: Optional[list[MotionLayer]] = None


# ─── Scene ──────────────────────────────────────────────────────────────────

class TransitionType(str, Enum):
    none = "none"
    fade = "fade"
    slide_left = "slideLeft"
    slide_right = "slideRight"
    slide_up = "slideUp"
    slide_down = "slideDown"
    dissolve = "dissolve"
    wipe = "wipe"


class SceneTransition(BaseModel):
    type: TransitionType = TransitionType.none
    duration: float = 0.5


class MotionScene(BaseModel):
    id: str
    name: str
    duration: float = 5
    bgColor: str = "#000000"
    layers: list[MotionLayer] = Field(default_factory=list)
    transition: Optional[SceneTransition] = None


# ─── Project ────────────────────────────────────────────────────────────────

class MotionProject(BaseModel):
    id: str
    name: str
    canvasWidth: int = 1920
    canvasHeight: int = 1080
    fps: int = 30
    scenes: list[MotionScene] = Field(default_factory=list)
    createdAt: str = ""
    updatedAt: str = ""


# ─── API Request/Response ───────────────────────────────────────────────────

class MotionPromptRequest(BaseModel):
    prompt: str
    canvas_width: int = 1920
    canvas_height: int = 1080
    duration: float = 5
    style: Optional[str] = None  # e.g. "minimal", "bold", "corporate"


class MotionEditRequest(BaseModel):
    instruction: str
    scene_id: str


class AnimationPreset(BaseModel):
    id: str
    name: str
    category: str  # enter, exit, emphasis, transition
    description: str
    keyframes: dict[str, list[Keyframe]]
    duration: float = 1.0

File: frontend/src/types/motion.ts
// ─── Easing ───────────────────────────────────────────────────────────────────

export type CubicBezier = [number, number, number, number];

export type EasingType =
  | 'linear'
  | 'easeIn'
  | 'easeOut'
  | 'easeInOut'
  | 'spring'
  | 'bounce'
  | CubicBezier;

// ─── Keyframes ────────────────────────────────────────────────────────────────

export interface Keyframe {
  id: string;
  time: number; // seconds
  value: number | string;
  easing: EasingType;
}

/** A single animatable property path, e.g. "x", "opacity", "fill" */
export type AnimatableProperty =
  | 'x'
  | 'y'
  | 'width'
  | 'height'
  | 'rotation'
  | 'opacity'
  | 'scaleX'
  | 'scaleY'
  | 'anchorX'
  | 'anchorY'
  | 'fill'
  | 'stroke'
  | 'strokeWidth'
  | 'borderRadius'
  | 'blur'
  | 'shadowX'
  | 'shadowY'
  | 'shadowBlur'
  | 'shadowOpacity'
  | 'fontSize'
  | 'letterSpacing'
  | 'lineHeight'
  | 'trimStart'
  | 'trimEnd';

// ─── Layer Properties ─────────────────────────────────────────────────────────

export interface LayerTransform {
  x: number;
  y: number;
  width: number;
  height: number;
  rotation: number;
  opacity: number;
  scaleX: number;
  scaleY: number;
  anchorX: number; // 0-1 relative
  anchorY: number; // 0-1 relative
}

export interface LayerAppearance {
  fill: string;
  stroke: string;
  strokeWidth: number;
  borderRadius: number;
}

export interface LayerEffects {
  blur: number;
  shadowX: number;
  shadowY: number;
  shadowBlur: number;
  shadowColor: string;
  shadowOpacity: number;
}

export interface TextProperties {
  text: string;
  fontSize: number;
  fontFamily: string;
  fontWeight: number;
  letterSpacing: number;
  lineHeight: number;
  textAlign: 'left' | 'center' | 'right';
  color: string;
}

export type ShapeType = 'rectangle' | 'ellipse' | 'polygon' | 'line' | 'star';

export interface ShapeProperties {
  shapeType: ShapeType;
  points?: number; // polygon/star vertex count
}

export interface PathProperties {
  trimStart: number; // 0-1 for stroke draw-on
  trimEnd: number;
}

// ─── Layers ───────────────────────────────────────────────────────────────────

export type LayerType = 'text' | 'shape' | 'image' | 'video' | 'group' | 'svg';

export interface MotionLayer {
  id: string;
  name: string;
  type: LayerType;
  visible: boolean;
  locked: boolean;
  inPoint: number;  // when layer appears (seconds)
  outPoint: number; // when layer disappears (seconds)
  transform: LayerTransform;
  appearance: LayerAppearance;
  effects: LayerEffects;
  textProps?: TextProperties;
  shapeProps?: ShapeProperties;
  pathProps?: PathProperties;
  imageSrc?: string;
  videoSrc?: string;
  keyframes: Record<string, Keyframe[]>; // propertyPath → sorted keyframes
  children?: MotionLayer[]; // nested group children
}

// ─── Scenes ───────────────────────────────────────────────────────────────────

export type TransitionType = 'none' | 'fade' | 'slideLeft' | 'slideRight' | 'slideUp' | 'slideDown' | 'dissolve' | 'wipe';

export interface SceneTransition {
  type: TransitionType;
  duration: number; // seconds
}

export interface MotionScene {
  id: string;
  name: string;
  duration: number; // seconds
  bgColor: string;
  layers: MotionLayer[];
  transition?: SceneTransition;
}

// ─── Project ──────────────────────────────────────────────────────────────────

export interface MotionProject {
  id: string;
  name: string;
  canvasWidth: number;
  canvasHeight: number;
  fps: number;
  scenes: MotionScene[];
  createdAt: string;
  updatedAt: string;
}

// ─── Presets ──────────────────────────────────────────────────────────────────

export interface AnimationPreset {
  id: string;
  name: string;
  category: 'enter' | 'exit' | 'emphasis' | 'transition';
  description: string;
  keyframes: Record<string, Keyframe[]>; // property -> keyframes (relative time 0-1)
  duration: number; // default duration in seconds
}

// ─── Editor State ─────────────────────────────────────────────────────────────

export interface PlaybackState {
  isPlaying: boolean;
  currentTime: number;
  loop: boolean;
}

export interface SelectionState {
  layerIds: string[];
  keyframeIds: string[];
}
